### PR TITLE
Currencycom markets

### PR DESCRIPTION
**Pull request by samgermain**
Unifies the markets of currency.com to match other exchanges

-------------------------------------

Some variables were shuffled around, but if I remove all the added variables (test by switching to second last commit) and run the cli `diff` command on the output of this branch and master I get

```
< SHIB/USD_LEVERAGE |                                 SHIB/USD_LEVERAGE |                                          SHIB |   USD |           SHIB |     USD | margin | false |   true |   true |     {"amount":10000,"price":1e-8} |   {"amount":{"min":10000,"max":100000000},"price":{},"cost":{"min":0.3317}} |  0.00075 | 0.00075
---
> SHIB/USD_LEVERAGE |                                 SHIB/USD_LEVERAGE |                                          SHIB |   USD |           SHIB |     USD | margin | false |   true |   true |     {"amount":10000,"price":1e-8} |  {"amount":{"min":10000,"max":100000000},"price":{},"cost":{"min":0.33175}} |  0.00075 | 0.00075
```

Because `cost.min` in `0.03317` on master and `0.33175` on this branch, do you know what would make that difference?

### CODE REVIEW

**Comment by kroitor** [comment]
We should avoid all floating-point math like `Math.xxx` in all parsers, including the market parser, and stick to Precise instead.

**Comment by samgermain** [comment]
> We should avoid all floating-point math like `Math.xxx` in all parsers, including the market parser, and stick to `Precise` instead.

It doesn't look like there's a `Precise.pow` or a `Precise.log10`

I didn't make any edits to the math that was in there, I just shifted things around

**Comment by kroitor** [comment]
> It doesn't look like there's a `Precise.pow` or a `Precise.log10`

We don't need them in general, cause we can do `Precise.stringMultiply ('1', '1e10');` etc.

> I didn't make any edits to the math that was in there, I just shifted things around

Yeah, i know, I just proposed to fix that along the way. If anything, you can leave that as is, i'll refix it later.

**Comment by samgermain** [comment]
> > It doesn't look like there's a `Precise.pow` or a `Precise.log10`
> 
> We don't need them in general, cause we can do `Precise.stringMul ('15, '1.7e10');`, `Precise.stringDiv`, etc.
> 
> > I didn't make any edits to the math that was in there, I just shifted things around
> 
> Yeah, i know, I just proposed to fix that along the way. If anything, you can leave that as is, i'll refix it later.

How would the power math work, seeing  as the exponent is dynamic? Also why are we even calculating 1 to the power of something, isn't that always 1?

**Comment by kroitor** [comment]
> How would the power math work, seeing as the exponent is dynamic? 

```JavaScript
            let precisionAmount = 1 / Math.pow (1, this.safeInteger (market, 'baseAssetPrecision'));
```
↓
```
            const baseAssetPrecision = this.safeString (market, 'baseAssetPrecision');
            const precisionAmount = this.parseNumber ('1e-' + baseAssetPrecision);
```

> Also why are we even calculating 1 to the power of something, isn't that always 1?

I was just typing quickly, you can `0.1e10` if you want.

**Comment by kroitor** [comment]
> How would the power math work, seeing as the exponent is dynamic? Also why are we even calculating 1 to the power of something, isn't that always 1?

```bash
~/ccxt $ node examples/js/cli ftx parseNumber 1e4
10000
```

**Comment by samgermain** [comment]
> > How would the power math work, seeing as the exponent is dynamic?
> 
> ```js
>             let precisionAmount = 1 / Math.pow (1, this.safeInteger (market, 'baseAssetPrecision'));
> ```
> 
> ↓
> 
> ```js
>             const baseAssetPrecision = this.safeString (market, 'baseAssetPrecision');
>             const precisionAmount = this.parseNumber ('1e-' + baseAssetPrecision);
> ```
> 
> > Also why are we even calculating 1 to the power of something, isn't that always 1?
> 
> I was just typing quickly, you can `0.1e10` if you want.

What about `log10`

**Comment by frosty00** [comment]
@samgermain you can parse a number using `new Precise` and inside you can find out how big it is by calling `.reduce` and then `.decimals`

so like so

```
exchangeReplyNumber = this.safeString (reply, 'amount')
precise = new Precise (exchangeReplyNumber)
precise.reduce ()
console.log (precise.decimals)
```

**Comment by samgermain** [comment]
> @samgermain you can parse a number using `new Precise` and inside you can find out how big it is by calling `.reduce` and then `.decimals`
> 
> so like so
> 
> ```
> exchangeReplyNumber = this.safeString (reply, 'amount')
> precise = new Precise (exchangeReplyNumber)
> precise.reduce ()
> console.log (precise.decimals)
> ```

It's been a while since I've broken these things down, is this correct?

```
log10(y) = x
10^x = y
x = 10^1/y
 ```

**Comment by samgermain** [comment]
> > How would the power math work, seeing as the exponent is dynamic?
> 
> ```js
>             let precisionAmount = 1 / Math.pow (1, this.safeInteger (market, 'baseAssetPrecision'));
> ```
> 
> ↓
> 
> ```js
>             const baseAssetPrecision = this.safeString (market, 'baseAssetPrecision');
>             const precisionAmount = this.parseNumber ('1e-' + baseAssetPrecision);
> ```
> 
> > Also why are we even calculating 1 to the power of something, isn't that always 1?
> 
> I was just typing quickly, you can `0.1e10` if you want.

What are regular powers, like `Math.pow (10, -precisionAmount)`

**Comment by kroitor** [comment]
@samgermain 

> What are regular powers, like `Math.pow (10, -precisionAmount)`

I'm not sure i fully understood the question, can you elaborate a bit?

We could simply:

```JavaScript
            const baseAssetPrecision = this.safeString (market, 'baseAssetPrecision');
            const precisionAmount = this.parseNumber ('1e-' + baseAssetPrecision);
```

**Comment by kroitor** [comment]
Related: #10931

**Comment by frosty00** [comment]
> @samgermain
> 
> > What are regular powers, like `Math.pow (10, -precisionAmount)`
> 
> I'm not sure i fully understood the question, can you elaborate a bit?
> 
> We could simply:
> 
> ```js
>             const baseAssetPrecision = this.safeString (market, 'baseAssetPrecision');
>             const precisionAmount = this.parseNumber ('1e-' + baseAssetPrecision);
> ```

we should use `parsePrecision` here

File: js/currencycom.js [hunk 1]
```diff
@@ -295,9 +303,9 @@ module.exports = class currencycom extends Exchange {
                     limitPriceMax = maxPrice;
                 }
             }
-            let precisionAmount = 1 / Math.pow (1, this.safeInteger (market, 'baseAssetPrecision'));
+            let precisionAmount = Precise.stringDiv ('1', this.parsePrecision (this.safeString (market, 'baseAssetPrecision')));
             let limitAmount = {
-                'min': Math.pow (10, -precisionAmount),
+                'min': this.parsePrecision ('-' + precisionAmount),
```
Comment (kroitor) [inline js/currencycom.js:308]: I think we don't need to prepend the negative sign here, cause it will be negated in parsePrecision.

**Pull request merged**